implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright 2024 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.registry.otlp;
+
+import io.micrometer.common.lang.Nullable;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.*;
+import io.micrometer.core.instrument.config.NamingConvention;
+import io.micrometer.core.instrument.distribution.CountAtBucket;
+import io.micrometer.core.instrument.distribution.HistogramSnapshot;
+import io.micrometer.core.instrument.distribution.HistogramSupport;
+import io.micrometer.core.instrument.distribution.ValueAtPercentile;
+import io.micrometer.core.instrument.util.TimeUtils;
+import io.opentelemetry.proto.common.v1.AnyValue;
+import io.opentelemetry.proto.common.v1.KeyValue;
+import io.opentelemetry.proto.metrics.v1.*;
+import io.opentelemetry.proto.metrics.v1.Metric.DataCase;
+
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+import java.util.function.DoubleSupplier;
+import java.util.stream.Collectors;
+
+/**
+ * A bridge for converting Micrometer meters to OTLP metrics.
+ */
+class OtlpMetricConverter {
+
+    private final Clock clock;
+
+    private final Duration step;
+
+    private final AggregationTemporality aggregationTemporality;
+
+    private final io.opentelemetry.proto.metrics.v1.AggregationTemporality otlpAggregationTemporality;
+
+    private final TimeUnit baseTimeUnit;
+
+    private final NamingConvention namingConvention;
+
+    private final Map<MetricMetaData, Metric.Builder> metricTypeBuilderMap = new HashMap<>();
+
+    private final long deltaTimeUnixNano;
+
+    OtlpMetricConverter(Clock clock, Duration step, TimeUnit baseTimeUnit,
+            AggregationTemporality aggregationTemporality, NamingConvention namingConvention) {
+        this.clock = clock;
+        this.step = step;
+        this.aggregationTemporality = aggregationTemporality;
+        this.otlpAggregationTemporality = AggregationTemporality.toOtlpAggregationTemporality(aggregationTemporality);
+        this.baseTimeUnit = baseTimeUnit;
+        this.namingConvention = namingConvention;
+        this.deltaTimeUnixNano = (clock.wallTime() / step.toMillis()) * step.toNanos();
```

Comment:
Isn't this (in combination with where the constructor is used in `publish`) changing from setting the delta time once per publish to setting it once per batch? I guess we don't have a test that would catch such a change.
Was the change intentional? It seems unrelated to the main purpose of the pull request.

Reply:
Yeah, this was intentional to this per batch reason being, in normal scenarios where the meters are all read and marshaled into OTLP protocol completed within the step this would mean no change except that computation of delatTimeUnixNano happens multiple times (once per batch). NOTE: Irrespective of how many computations happen the value is the same.

In case, there are too many and all of them are not read within the same step interval (like half are read and some spill over), the value would have rolled over to the next step, and re-computing per batch would at least set the right timestamp for that step. But maybe the right thing would be to re-compute for every meter. That looks sub-optimal to me given that if this scenario happens already the metrics are flawed and users have to re-configure their steps to accommodate publishing all the meters. What do you think?

Reply:
I think it's fine to make the change, but it should be called out as a change in behavior that was made. This is both for review purposes and tracking/announcing changes to users. It's not clear from the title or description that this would be changed as part of this.

Reply:
I have also updated the description of PR to reflect the same.